### ISSUE TITLE
dimensioning bug?

### ISSUE THREAD

**Issue by jtatusko** [huggingface/pytorch-openai-transformer-lm#2]
https://github.com/huggingface/pytorch-openai-transformer-lm/blob/master/model_py.py#L77-L84

The reference implementation and paper uses tensorflow, in which channel dimension is represented last. But in pytorch, the channel dimension is represented as dim=1 after the batch. So does the equation need to be reversed? Is this done somewhere else in the code?

so in pytorch it should look like: matmul(v, matmul(k.t(), q))

example (tensorflow):
```
q size [attn_depth=3, feature_depth=2]
k size [attn_depth=3, feature_depth=2]
v size [attn_depth=3, feature_depth=2]

weights = matmul(q, k.t()) -> [3, 3]
result = matmul(weights, v) -> [3, 2]
```

example (pytorch) with error:
```
q size [2, 3]
k size [2, 3]
v size [2, 3]

weights = matmul(q, k.t()) -> [2, 2]
result = matmul(weights, v) -> [2, 3]
```
Notice the dimensionality of the weights has been over-reduced [2,2] instead of [3,3]

example (pytorch) with correction:
```
q size [2, 3]
k size [2, 3]
v size [2, 3]

weights = matmul(k.t(), q) -> [3, 3]
result = matmul(v, weights) -> [2, 3]
```
Apologies if this is handled correctly somewhere, limited on time atm for a thorough read.

**Comment by thomwolf** [comment]
Are you talking about this permutation on the keys ? https://github.com/huggingface/pytorch-openai-transformer-lm/blob/master/model_py.py#L94 (called from here: https://github.com/huggingface/pytorch-openai-transformer-lm/blob/master/model_py.py#L103)

**Comment by jtatusko** [comment]
oops, I see why you transpose the key, it's not transposed in the equation. minor nitpick would be to just transpose it in _attn for readability (equation will appear as it does in the paper). And yeah, if your x's dimensions look like [batch, length, channels] then the code looks correct to me :thumbsup:.

Thanks for open sourcing this

**Comment by thomwolf** [comment]
You are welcome!
